internal/controllers/machine/machine_controller.go

```diff
@@ -670,8 +670,18 @@ func (r *Reconciler) drainNode(ctx context.Context, cluster *clusterv1.Cluster,
 	}
 
 	if noderefutil.IsNodeUnreachable(node) {
-		// When the node is unreachable and some pods are not evicted for as long as this timeout, we ignore them.
-		drainer.SkipWaitForDeleteTimeoutSeconds = 60 * 5 // 5 minutes
+		// Kubelet is unreachable, pods will never disappear.
+
+		// SkipWaitForDeleteTimeoutSeconds ensures the drain completes
+		// even if pod objects are not deleted.
+		drainer.SkipWaitForDeleteTimeoutSeconds = 1
+
+		// kube-apiserver sets the `deletionTimestamp` to a future date computed using the grace period.
+		// We are effectively waiting for GracePeriodSeconds + SkipWaitForDeleteTimeoutSeconds.
+		// Override the grace period of pods to reduce the time needed to skip them.
+		drainer.GracePeriodSeconds = 1
+
+		log.V(5).Info("Node is unreachable, draining will ignore gracePeriod. PDBs are still honored.")
```

Comment:
your assumption is correct. But not because we don't honour PDBs. We still make the evict API request which account for pdbs. But because of https://github.com/kubernetes/kubernetes/pull/94381 and https://github.com/kubernetes/kubernetes/pull/91342 the request legitimately succeeds as the unhealthy pods do not decrease the disruption budget anymore. Before those fixes drain might block forever while deleting an unreachable node even though the pods and nodes were doomed.

Reply:
Pods on unreachable nodes (ie ready: unknown) are automatically marked as unhealthy? (I guess by setting the pod ready condition to unknown)

Reply:
I couldn't find it explicitly called out in the docs.

But looks like at the same time when the Ready condition of the Node is flagged as Unknown the Ready condition of the Pod is set to False

```
# Node:
  - lastHeartbeatTime: "2024-06-14T09:25:35Z"
    lastTransitionTime: "2024-06-14T09:27:48Z"
    message: Kubelet stopped posting node status.
    reason: NodeStatusUnknown
    status: Unknown
    type: Ready

# Pod
  - lastProbeTime: null
    lastTransitionTime: "2024-06-14T09:27:48Z"
    status: "False"
    type: Ready
```

Reply:
For reference. This seems to be the code that marks Pods as not ready when the Node is not ready: https://github.com/kubernetes/kubernetes/blob/6d0ac8c561a7ac66c21e4ee7bd1976c2ecedbf32/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L1151

Reply:
Back to the original point

> your assumption is correct. But not because we don't honour PDBs.

Thx, I took another look at the drain helper and eviction and yeah this is obviously true :joy:

Reply:
also for record the node life cycle controller automatically starts evicting all the pods when a kubelet is unreachable for longer than N https://kubernetes.io/docs/concepts/architecture/nodes/#node-controller